File: pythonStudy/BOJ/1016 제곱ㄴㄴ수.py
# 1016 제곱 ㄴㄴ수
input = __import__('sys').stdin.readline

LMT = 1000000
sieve = [True]*(LMT+1)
Mi, Ma = map(int, input().split())
rng = Ma - Mi + 1
iterMax = int(Ma**0.5)
for i in range(2, iterMax+1):
    if i != 2 and i%2==0: continue
    if i != 3 and i%3==0: continue
    if i != 5 and i%5==0: continue
    if i != 7 and i%7==0: continue
    jg = i*i
    for mult in range(Mi//jg, (Ma//jg)+1):
        if (mult*jg < Mi): continue
        if (Ma < mult * jg): continue
        sieve[mult*jg - Mi] = False
print(sum(sieve[0:rng]))

File: C++/Solved/BOJ/1016 제곱ㄴㄴ수.cpp
#include <bits/stdc++.h>
using namespace std;
using ll = long long;
// 1016 제곱 ㄴㄴ수
#define FOR(i, n, m) for(ll i=(n); i<(m); i++)
constexpr int LMT = 1000000;

vector<bool> sieve(LMT + 1, true);
ll Mi, Ma;
int main(void) {
	ios::sync_with_stdio(0);
	cin.tie(0);

	cin >> Mi >> Ma;
	int rng = Ma - Mi + 1;
	ll iterMax = ll(sqrt(Ma));
	FOR(i, 2, iterMax + 1) {
		if (i != 2 && i % 2 == 0) continue;
		if (i != 3 && i % 3 == 0) continue;
		if (i != 5 && i % 5 == 0) continue;
		if (i != 7 && i % 7 == 0) continue;
		ll jg = i * i;
		FOR(mult, Mi / jg, (Ma / jg) + 1) {
			if (mult * jg < Mi) continue;
			if (Ma < mult * jg) continue;
			sieve[mult * jg - Mi] = false;
		}
	}
	int sm = 0;
	FOR(i, 0, rng) if (sieve[i]) sm++;
	cout << sm;
}